src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.mesos.state.StateStore;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides the core functionality of the automatic repair mechanisms.
+ */
+public class RepairScheduler {
+    private final Log log = LogFactory.getLog(getClass());
+
+    private final String targetConfigName;
+    private final StateStore stateStore;
+    private final OfferAccepter offerAccepter;
+    private final TaskFailureListener failureListener;
+    private final RepairOfferRequirementProvider offerReqProvider;
+    private final FailureMonitor failureMonitor;
+    private final LaunchConstrainer launchConstrainer;
+    private final AtomicReference<RepairStatus> repairStatusRef;
+
+    public RepairScheduler(
+            String targetConfigName,
+            StateStore stateStore,
+            TaskFailureListener failureListener,
+            RepairOfferRequirementProvider offerReqProvider,
+            OfferAccepter offerAccepter,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor,
+            AtomicReference<RepairStatus> repairStatusRef) {
+        this.targetConfigName = targetConfigName;
+        this.stateStore = stateStore;
+        this.offerReqProvider = offerReqProvider;
+        this.offerAccepter = offerAccepter;
+        this.failureMonitor = failureMonitor;
+        this.launchConstrainer = launchConstrainer;
+        this.repairStatusRef = repairStatusRef;
+        this.failureListener = failureListener;
+    }
+
+    public List<OfferID> resourceOffers(SchedulerDriver driver, List<Offer> offers, Block block)
+            throws Exception {
+        List<OfferID> acceptedOffers = new ArrayList<>();
+        List<TaskInfo> terminatedTasks = new ArrayList<>(getTerminatedTasks(block));
+
+        // Compute the stopped and failed pool for UI
+        RepairStatus oldStatus = repairStatusRef.get();
+        if (oldStatus == null) {
+            oldStatus = new RepairStatus(Collections.emptyList(), Collections.emptyList());
+        }
+        List<String> oldStopped = oldStatus.getStopped();
+        List<String> newStopped = new ArrayList<>(terminatedTasks.stream()
+                .filter(failureMonitor::hasFailed)
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList()));
+        newStopped.addAll(oldStopped);
+        //TODO (ghartman) possibly write this list down in ZK, to ensure the UI doesn't change upon restart
+        newStopped = newStopped.stream().distinct().collect(Collectors.toList());
+        List<String> newFailed = terminatedTasks.stream()
+                .filter(it -> !failureMonitor.hasFailed(it))
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList());
+
+        Optional<OfferRequirement> offerReq = Optional.empty();
+        boolean replacingStopped = false;
+
+        if (terminatedTasks.size() > 0) {
```

Comment:
Logging terminology in this `if / else` block is not consistent with the terms used in `RepairStatus`.  Use of the terms `stopped` and `failed` would be clearer.